components/region_cache_memory_engine/src/write_batch.rs

```diff
@@ -58,8 +58,14 @@ impl RangeCacheWriteBatch {
     }
 
     fn write_impl(&mut self, seq: u64) -> Result<()> {
+        self.engine.handle_pending_load();
         let (engine, filtered_keys) = {
-            let core = self.engine.core().read().unwrap();
+            let mut core = self.engine.core().write().unwrap();
```

Comment:
write() would make apply threads wait with each other when calls write_opt. To me here it must be able to run parallelly. Otherwise this would make apply further unscalable.